### DISCUSSION TITLE
[Question] How to show different Text when transcluded or not

### DISCUSSION THREAD

**Discussion by Parlazo** [TiddlyWiki/TiddlyWiki5#5687] [Q&A]
How can I enable or disable some parts of a Tiddler depending if it is transcluded or not?

The idea would be some thing like

`<<IfTranscluded TranscludedText:'' NotTranscludedText:''>>`
or
`<<IfTranscluded TranscludedText:'I am transcluded' NotTranscludedText:'I am not transcluded'>>`

I did try now a lot with any thing i could guess from regex against <<transcluded>> and <<currentTiddler>> but some how it did not work in a list ( `((?:[\+|\-|\*|\/|\^|\{|\}|\[|\]|\(|\)|\\|\$|\.|\||\?]*))\{(<<currentTiddler>>\?)\|(\1)(\2)\|(\1).*?\|(\1).*?\|(\1).*?\}` )
and <<qualified {{!!title}}>> revealing when match to <<qualified {{spaced Title!!title}}>>

may i got all wrong and honestly the whole concept looks confusing to me out of what i get in the docs plugin.

has any one a idea?

**Comment by Jermolene** [comment]
Hi @Parlazo

Tiddlers are always displayed via transclusion, but a simple test that can be used within a tiddler in the story river to check whether it has been transcluded is to compare the variable "storyTiddler" with the variable "currentTiddler". It's not foolproof because it won't detect transclusions that don't change the current tiddler, but it will be sufficient for many cases:

```
<$list filter="[<currentTiddler>match<storyTiddler>]" emptyMessage="This tiddler is transcluded">
This tiddler is not transcluded
</$list>
```

**Reply by Parlazo** [reply]
Based on your approach with the match-filter i had a new idea

Macro Tiddler Title: `TransTexter`
Macro Tiddler Tag: `$:/tags/Macro`
```
\define TransTexterInfo(Comment:'')
<!--
  Macro: Yes

  Title: TransTexter

  Tag´s: $:/tags/Macro

  Usage: 
    #''Simple Transclusion:''
    #*<<TransTexter NoTransclusionText:""" "[[I]] 'am' not {transcluded} |text|." """ TransclusionText:""" "[[I]] 'am' {transcluded} |text|." """>>
    #''Macro Call transclusion:''
    #*<$macrocall $name="TransTexter" NoTransclusionText="""//"[[I]] 'am' not {transcluded} |text|."//""" TransclusionText="""//"[[I]] 'am' {transcluded} |text|."//"""/>
    #''Transclusion sans parameter:''
    #*<<TransTexter>>

  Infos:
    The macro will show different text when a Tiddler is transcluded or not.
    It will deactivate [[links]] and {{transclusion}}
    If needed you can activate this by editing the macro.
      Replace NoTransTextEntityFix by NoTransclusionText
      Replace TransTextEntityFix by TransclusionText

  Notes:
    If you use the macro with no parameter, the default output will look empty, but it is a hidden HTML comment.
    You can fix this by replacing the two strings ...
      <!-- I am not transcluded default hidden text. -->
      <!-- I am transcluded default hidden text. -->
    ... with what you like as your default text.
-->
\end

\define RegExTransclusionToSimpleString() \||\{|\}
\define RegExParameterToDQString() \".*?
\define RegExParameterToApString() \'.*?
\define RegExParameterToLbrackString() \[.*?
\define RegExParameterToLbraceString() \{.*?

\define TransTexter(
NoTransclusionText:'<!-- I am not transcluded default hidden text. -->'
TransclusionText:'<!-- I am transcluded default hidden text. -->'
)
<!--
  Replace all difficult Characters in NoTransclusionText:
  " make doublequotes to entity
  ' make quotes to entity
  [ make left square brackets to entity, to prevent Links
  { make left curly brackets to entity, prevent transclusion
-->
<$list filter='[<__NoTransclusionText__>search-replace:gi:regexp<RegExParameterToDQString>,[&quot;]] +[search-replace:gi:regexp<RegExParameterToApString>,[&apos;]] +[search-replace:gi:regexp<RegExParameterToLbrackString>,[&#91;]]
+[search-replace:gi:regexp<RegExParameterToLbraceString>,[&#123;]]' variable='NoTransTextEntityFix'>
<!--
  Replace all difficult Characters in TransclusionText:
  " make doublequotes to entity
  ' make quotes to entity
  [ make left square brackets to entity, to prevent Links
  { make left curly brackets to entity, prevent transclusion
-->
<$list filter='[<__TransclusionText__>search-replace:gi:regexp<RegExParameterToDQString>,[&quot;]] +[search-replace:gi:regexp<RegExParameterToApString>,[&apos;]] +[search-replace:gi:regexp<RegExParameterToLbrackString>,[&#91;]]
+[search-replace:gi:regexp<RegExParameterToLbraceString>,[&#123;]]' variable='TransTextEntityFix'>
<!--
  Replace all difficult Characters in <<transclusion>>:
  Remove "{" or "|" or "}"
  Clear first appearance of <<currentTiddler>>
  Replace second appearance of <<currentTiddler>> with a entity fixed NoTransclusionText
-->
<$list filter="[<transclusion>search-replace:gi:regexp<RegExTransclusionToSimpleString>,[]search-replace::<currentTiddler>,[]search-replace::<currentTiddler>,<NoTransTextEntityFix>]" variable='NTT'>
<!--
  Check for leftover from <<transclusion>> which is not <<currentTiddler>>:
  If we have removed <<currentTiddler>> twice, in the previous step, then we show what we have replaced it with.
  If we have not removed <<currentTiddler>> twice, in the previous step, then we show a entity fixed TransclusionText.
-->
<$list filter="[<NTT>match<NoTransTextEntityFix>]" emptyMessage='<<TransTextEntityFix>>' variable='TransOut'>
<<TransOut>>
</$list>
<!-- TRANSPARENTOR -->
</$list>
<!-- TRANSPARENTOR -->
</$list>
<!-- TRANSPARENTOR -->
</$list>
\end
```

Usage in a Tiddler which is shown as Tiddler and then gets transcluded and shown in an other Tiddler at the same time.
```
    #''Simple Transclusion:''
    #*<<TransTexter NoTransclusionText:""" "[[I]] 'am' not {transcluded} |text|." """ TransclusionText:""" "[[I]] 'am' {transcluded} |text|." """>>
    #''Macro Call transclusion:''
    #*<$macrocall $name="TransTexter" NoTransclusionText="""//"[[I]] 'am' not {transcluded} |text|."//""" TransclusionText="""//"[[I]] 'am' {transcluded} |text|."//"""/>
    #''Transclusion sans parameter:''
    #*<<TransTexter>>
```

It will show in the original Tiddler an other text as it will show in the transcluding Tiddler.

There are some filters on the string, which is given as a parameter to the macro, which deactivate [[links]] or {{transclusion}}. This can get undone easily, as you like. If no parameter is given it will return an empty string, this is so because in that way you can show some text only when transcluded or only when not transcluded, or different strings when transcluded or not transcluded.

Its my first TiddlyWiki macro ever.
Sorry my bad english and sorry for the crazy code style.

May some one can use it later or beautify the code and some how include it into the TiddlyWiki core, because i think such a functionality makes sense.

Thanks to every one!

**Comment by Parlazo** [comment]
Just in case any one is interested, this was an older approach.
It is HardCoded in the Tiddler which will get transcluded later.
It will show the emptyMessage if the containing Tiddler is transcluded.
It will show the text in the list if the Tiddler is not transcluded.

The drawback of this approach is:
You have to put this into every Tiddler where you plan to use it and you have to use it on every text where you want to apply it.
The shown or not shown text is hardcoded and the title of the tiddler which is containing it is hardcoded too.

```
<!-- If transcluded. -->
! <$list
filter='[<transclusion>prefix[{HardcodedActualTiddlerTitleHere|]]'
emptyMessage=' [[HardcodedActualTiddlerTitleHere|HardcodedActualTiddlerTitleHere]]'>
</$list>

<!-- If not transcluded. -->
!! <$list
filter='[<transclusion>prefix[{HardcodedActualTiddlerTitleHere|]]'
emptyMessage=''>
TextIfNotTranscluded
</$list>
```

May some one just wants to use such a functionality once, so this is maybe the shorter approach if you are OK with the drawbacks.
Otherwise the Macro above is the way to go, as far as i can see at the moment.

Thanks to every one.
